### PR TITLE
Feature/680 add metadata

### PR DESCRIPTION
**Pull request by JulianFeinauer**
Draft of current Implementation for feedback.

Fixes: #680

### CODE REVIEW

File: model/things/src/main/java/org/eclipse/ditto/model/things/ImmutableThing.java
Comment (thjaeckle) [inline model/things/src/main/java/org/eclipse/ditto/model/things/ImmutableThing.java:550]: `metadata` is still missing in the `toString()` of this class

File: model/things/src/main/java/org/eclipse/ditto/model/things/ImmutableThingFromCopyBuilder.java
Comment (thjaeckle) [inline model/things/src/main/java/org/eclipse/ditto/model/things/ImmutableThingFromCopyBuilder.java:703]: What is this method for - not overriding an interface method?

File: model/things/src/main/java/org/eclipse/ditto/model/things/ImmutableThingFromScratchBuilder.java
Comment (thjaeckle) [inline model/things/src/main/java/org/eclipse/ditto/model/things/ImmutableThingFromScratchBuilder.java:341]: the params should be `final` (its our style for parameters)

File: model/things/src/main/java/org/eclipse/ditto/model/things/ImmutableThingFromCopyBuilder.java
Comment (thjaeckle) [inline model/things/src/main/java/org/eclipse/ditto/model/things/ImmutableThingFromCopyBuilder.java:698]: the params should be `final` (its our style for parameters)

File: model/things/src/main/java/org/eclipse/ditto/model/things/MetadataBuilder.java
Comment (thjaeckle) [inline model/things/src/main/java/org/eclipse/ditto/model/things/MetadataBuilder.java:196]: Did you not want to add a `Metadata` specific Javadoc here?

File: model/things/src/main/java/org/eclipse/ditto/model/things/MetadataBuilder.java
Comment (thjaeckle) [inline model/things/src/main/java/org/eclipse/ditto/model/things/MetadataBuilder.java:200]: Did you not want to add a `Metadata` specific Javadoc here?

File: model/things/src/main/java/org/eclipse/ditto/model/things/MetadataModelFactory.java
Comment (thjaeckle) [inline model/things/src/main/java/org/eclipse/ditto/model/things/MetadataModelFactory.java:28]: Please add Javadoc for a public class in the model

File: model/things/src/main/java/org/eclipse/ditto/model/things/Thing.java
Comment (thjaeckle) [inline model/things/src/main/java/org/eclipse/ditto/model/things/Thing.java:662]: I would move this up below the "_modified" definition to have them all in one place.

File: services/things/persistence/src/main/java/org/eclipse/ditto/services/things/persistence/actors/strategies/events/AbstractThingEventStrategy.java
Comment (thjaeckle) [inline services/things/persistence/src/main/java/org/eclipse/ditto/services/things/persistence/actors/strategies/events/AbstractThingEventStrategy.java:54]: I think these changes can be removed, right?
We would need a mechanism to generically add metadata provided in the headers (with a defined prefix) to the thing's `Metadata`.

File: model/things/src/main/java/org/eclipse/ditto/model/things/ImmutableMetadataBuilder.java
Comment (thjaeckle) [inline model/things/src/main/java/org/eclipse/ditto/model/things/ImmutableMetadataBuilder.java:2]: Please change the year to 2020

File: model/things/src/main/java/org/eclipse/ditto/model/things/ImmutableThingFromCopyBuilder.java
Comment (JulianFeinauer) [reply model/things/src/main/java/org/eclipse/ditto/model/things/ImmutableThingFromCopyBuilder.java:703]: @thjaeckle I was unsure about this one. This method is a shorthand and is called in the `of(JsonObject)` method as FP. I dont know if we shuld then move it up to the interface. It could even be private here, in theory or we just call the other method `(JsonPointer, Metadata)` and pass empty Pointer explicitly. What would you prefer?

File: services/things/persistence/src/main/java/org/eclipse/ditto/services/things/persistence/actors/strategies/events/AbstractThingEventStrategy.java
Comment (JulianFeinauer) [reply services/things/persistence/src/main/java/org/eclipse/ditto/services/things/persistence/actors/strategies/events/AbstractThingEventStrategy.java:54]: Yes, you are right, thats from me "playing around".

**Comment by JulianFeinauer** [comment]
Thank you for the review @thjaeckle I agree with all points and added more specific questions where needed. I will resolve these things in the next time and wait for another review. In the meantime I will try to take care of the ditto headers. Although we could also do it in a separate PR then things stay a bit simpler. WDYT?

File: model/things/src/main/java/org/eclipse/ditto/model/things/ImmutableThingFromCopyBuilder.java
Comment (thjaeckle) [reply model/things/src/main/java/org/eclipse/ditto/model/things/ImmutableThingFromCopyBuilder.java:703]: I would prefer passing the empty Pointer explicitly to the already existing interface method.

**Comment by thjaeckle** [comment]
> In the meantime I will try to take care of the ditto headers. Although we could also do it in a separate PR then things stay a bit simpler. WDYT?

I would add this to this PR as the `Metadata` is not usable without setting it via the ditto headers. :)

**Comment by JulianFeinauer** [comment]
@thjaeckle I fixed all reported issues and tried the first implementation fo "header parsing" and metadata creation from it as you suggest it in the issue.

File: model/things/src/main/java/org/eclipse/ditto/model/things/Thing.java
Comment (thjaeckle) [inline model/things/src/main/java/org/eclipse/ditto/model/things/Thing.java:578]: please add `@since 1.2.0` javadoc below `@return`

File: model/things/src/main/java/org/eclipse/ditto/model/things/MetadataModelFactory.java
Comment (thjaeckle) [inline model/things/src/main/java/org/eclipse/ditto/model/things/MetadataModelFactory.java:28]: please add `@since 1.2.0` javadoc

File: services/things/persistence/src/main/java/org/eclipse/ditto/services/things/persistence/actors/strategies/events/AbstractThingEventStrategy.java
Comment (thjaeckle) [inline services/things/persistence/src/main/java/org/eclipse/ditto/services/things/persistence/actors/strategies/events/AbstractThingEventStrategy.java:56]: variable can be `final`

File: services/things/persistence/src/main/java/org/eclipse/ditto/services/things/persistence/actors/strategies/events/AbstractThingEventStrategy.java
Comment (thjaeckle) [inline services/things/persistence/src/main/java/org/eclipse/ditto/services/things/persistence/actors/strategies/events/AbstractThingEventStrategy.java:65]: I'm afraid that it's not that easy.
In https://github.com/eclipse/ditto/issues/680#issuecomment-654165747 I mentioned different syntaxes for the `ditto-metadata:`:
* "ditto-metadata:issuedAt": "2020-06-09T14:29:00Z"
   * the idea of this syntax is that each JSON "leaf" which was updated in the `event` shall receive the metadata `"issuedAt": "2020-06-09T14:29:00Z"`
   * therefore, the `event.getResourcePath()` has to be taken into account (as base path of which part of the thing was changed - this will be `/` when the complete thing was modified) , combined with all pathes of the JSON leafes which can be found out via `event.getEntity(event.getImplementedSchemaVersion())`
   * a JsonObject has to be created for each JSON leaf containing the set metadata
* "ditto-metadata:/features/lamp/properties/color/r/issuedAt": "2020-06-09T14:14:00Z"
   * the idea of this syntax is to specify "overwrites" for the syntax above

For example:
You modify the feature properties of a "Lamp" feature in this Thing:
```json
{
  "thingId": "org.eclipse.ditto:my-lamp",
  "policyId": "org.eclipse.ditto:my-policy",
  "features": {
    "Lamp": { 
      "properties": {
        "on": true,
        "dimmer": 0.8
      }
    }
  }
}
```

You modify `"/features/Lamp/properties"` to:
```json
{
  "on": false,
  "dimmer": 0.0
}
```

The event looks like:
* `getResourcePath()` returns a JsonPointer with `"/features/Lamp/properties"`
*  `event.getEntity(event.getImplementedSchemaVersion())` returns the above JsonObject

If now the header `"ditto-metadata:issuedAt": "2020-06-09T14:29:00Z"` was set, we want to have the following thing JSON (incl. `_metadata`) as a result:
```json
{
  "thingId": "org.eclipse.ditto:my-lamp",
  "policyId": "org.eclipse.ditto:my-policy",
  "features": {
    "Lamp": { 
      "properties": {
        "on": false,
        "dimmer": 0.0
      }
    }
  },
  "_metadata": {
    "features": {
      "Lamp": { 
        "properties": {
          "on": {
            "issuedAt": "2020-06-09T14:29:00Z"
          },
          "dimmer": {
            "issuedAt": "2020-06-09T14:29:00Z"
          }
        }
      }
    }
  }
}
```

So the mechanism is a lot more complex and I think we should extract it to another class in order to be able to also better unit test it.

Ah, and the idea would be to **always** add the following metadata for all affected JSON leafes:
* `_revision`: is passed in via the parameter `revision`
* `_modified`: can be obtained by `event.getTimestamp().orElse(null)`

Comment (JulianFeinauer) [reply services/things/persistence/src/main/java/org/eclipse/ditto/services/things/persistence/actors/strategies/events/AbstractThingEventStrategy.java:65]: Thanks for the pointers. Indeed it was not meant as complete implementation but a first draft. But your comment helps me. I will try to extract that to another class and add some tests to get a better understanding :)

File: model/things/src/main/java/org/eclipse/ditto/model/things/ImmutableThingFromCopyBuilder.java
Comment (jufickel-b) [reply model/things/src/main/java/org/eclipse/ditto/model/things/ImmutableThingFromCopyBuilder.java]: On Thing level the Pointer is unused and  thus it should be removed IMHO.

File: model/things/src/test/java/org/eclipse/ditto/model/things/ImmutableThingTest.java
Comment (jufickel-b) [inline model/things/src/test/java/org/eclipse/ditto/model/things/ImmutableThingTest.java:1136]: I would prefer the following implementation of that method:

    final JsonObject json = KNOWN_THING_V2.toJson(JsonSchemaVersion.V_2, field -> true);

    final Optional<JsonObject> metadataOptional = json.getValue(Thing.JsonFields.METADATA);

    assertThat(metadataOptional).contains(JsonObject.newBuilder().set("_issuedAt", JsonValue.of(0)).build());

File: model/things/src/main/java/org/eclipse/ditto/model/things/NullMetadata.java
Comment (thjaeckle) [reply model/things/src/main/java/org/eclipse/ditto/model/things/NullMetadata.java]: the same is done for `Attributes` so I would keep it that way

**Review by thjaeckle — APPROVED** [review]
I did a review of the complete PR.
I renamed the header key from "metadata" to "put-metadata" to be more precise and explicit what happens with the header value. Also I think that we will in the future need some header like "get-metadata" in order to e.g. retrieve metadata when doing a HTTP GET request for e.g. a single feature or property (to include the metadata as HTTP response header).

Afterwards I merged the latest master onto the PR branch and resolved the conflicts we got and I addressed the still open topics addressed in this PR discussion done by @jufickel-b 

So my approval from technical/code side 👍 
This is a great addition to Ditto - much looking forward to the possibilities it opens.

**Comment by thjaeckle** [comment]
I marked PR as ready for review, so that it can make it into Ditto 1.2.0 - the documentation can be added in another PR.

**Pull request merged**
